### DISCUSSION TITLE
Support for updating docker images in jib maven plugin

### DISCUSSION THREAD

**Discussion by andrein** [renovatebot/renovate#12774] [Suggest an Idea]
### What would you like Renovate to be able to do?

Renovate should be able to update the base image used by the [jib maven plugin](https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin)

Example `pom.xml`:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
[...]
	<build>
		<plugins>
			<plugin>
				<groupId>com.google.cloud.tools</groupId>
				<artifactId>jib-maven-plugin</artifactId>
				<version>3.0.0</version>
				<configuration>
					<from>
						<image>openjdk:11.0.10-jre</image> <!-- this should be updated -->
					</from>
					<to>
						<image>myimage</image>
					</to>
				</configuration>
			</plugin>
		</plugins>
	</build>
</project>
```

### If you have any ideas on how this should be implemented, please tell us here.

I'm thinking of changing [depFromNode()](https://github.com/renovatebot/renovate/blob/main/lib/manager/maven/extract.ts#L40) to return `PackageDependency[] | null` and returning an extra dependency if depName is `com.google.cloud.tools:jib-maven-plugin`

### Is this a feature you are interested in implementing yourself?

Yes

**Comment by viceice** [comment]
As workaround you can use regex manager for now

**Comment by andrein** [comment]
I've thought of that, but I have about a hundred apps to keep up to date :) I'd rather implement this in renovate directly.

**Comment by viceice** [comment]
What about shared presets? we recommend to use them when you need same settings for multiple repos. 
We do so ourself. :wink:

**Comment by rarkins** [comment]
Is this (`plugin.configuration.from.image`) standardized in any way or essentially a proprietary?

**Comment by andrein** [comment]
> What about shared presets? we recommend to use them when you need same settings for multiple repos. We do so ourself. 😉

I think we could use that. If we add it to onboardingConfig, is it possible to tell renovate to update all the repos with the new config?

> Is this (`plugin.configuration.from.image`) standardized in any way or essentially a proprietary?

I don't think it is standardized. I've only seen it used by the jib plugin so far.

**Reply by rarkins** [reply]
In that case if it's a proprietary feature of one plugin then we don't want this in the core maven manager but instead in a preset. We recommend you:
1. Use a centralized preset for all your repos so that you can apply this type of change universally whenever you need
2. Set it up so that that custom preset is part of your onboardingConfig: https://docs.renovatebot.com/config-presets/#organization-level-presets

**Reply by andrein** [reply]
Is there a way to force renovate to re-onboard all projects so we don't have to update the old projects manually?

**Reply by rarkins** [reply]
Not currently. And what would you do about projects which had non standard config? Eg you presumably wouldn't want them overwritten with the onboardingConfig

**Reply by andrein** [reply]
We only have a handful of projects that have non-standard configs, but I get why it can't be done :)

**Comment by shurkanTwo** [comment]
@andrein Have you found a solution for this? We also want to implement the exact same use case. And also want to "re-onboard all projects so we don't have to update the old projects manually". :)

**Comment by matzegebbe** [comment]
You just have to remember that it is really wild west with new versions and tags and thus may be difficult to dedect if a new version is available https://docs.renovatebot.com/modules/versioning/#docker-versioning

> Docker image authors can use whatever tag they want, it's a "wild west". Docker tags don't always follow SemVer. This means that Renovate tries to accept and sort SemVer-like versions, but this won't always work.

for me that worked:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "gitlab>hwl/dxp/cicd/renovate"
  ],
  "regexManagers": [
    {
      "description": "maven-jib",
      "fileMatch": [
        "(^|/)pom.xml$"
      ],
      "matchStringsStrategy": "any",
      "matchStrings": [
        "<from>\\s*<image>\\s*(?<depName>.*?):(?<currentValue>.*?)@?(?<currentDigest>(sha256:[a-f0-9]+)?)\\s*<\/image>\\s*<\/from>",
        "<from>\\s*<image>\\s*(?<registryUrl>.*\\/)(?<depName>.*?):(?<currentValue>.*?)@?(?<currentDigest>(sha256:[a-f0-9]+)?)\\s*<\/image>\\s*<\/from>"
      ],
      "datasourceTemplate": "docker"
    }
  ]
}
```
https://regex101.com/r/JTOaGK/1
![image](https://user-images.githubusercontent.com/3482021/222484408-ff2f2e55-78be-4c49-a59a-1b6cb8ab2571.png)

result in renovate (debug)

```json
"regex": [
           {
             "deps": [
               {
                 "depName": "eclipse-temurin",
                 "currentValue": "18-jdk-alpine",
                 "currentDigest": "sha256:4711",
                 "datasource": "docker",
                 "replaceString": "<from>\n                        <image>\n                            eclipse-temurin:18-jdk-alpine@sha256:4711\n                        </image>\n                    </from>",
                 "updates": [
                   {
                     "updateType": "digest",
                     "newValue": "18-jdk-alpine",
                     "newDigest": "sha256:7f1d3de830105436d0f0f10b831da319812e52f845a8d92d0dc036e45eb1244b",
                     "branchName": "renovate/eclipse-temurin-18-jdk-alpine"
                   }
                 ],
                 "warnings": [],
                 "versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
                 "registryUrl": "https://index.docker.io",
                 "currentVersion": "18-jdk-alpine",
                 "fixedVersion": "18-jdk-alpine"
               }
             ],
             "matchStrings": [
               "<from>\\s*<image>\\s*(?<depName>.*?):(?<currentValue>.*?)@?(?<currentDigest>(sha256:[a-f0-9]+)?)\\s*</image>\\s*</from>",
               "<from>\\s*<image>\\s*(?<registryUrl>.*\\/)(?<depName>.*?):(?<currentValue>.*?)@?(?<currentDigest>(sha256:[a-f0-9]+)?)\\s*</image>\\s*</from>"
             ],
             "matchStringsStrategy": "any",
             "datasourceTemplate": "docker",
             "packageFile": "pom.xml"
           }
         ]
```

**Comment by jackson-chris** [comment]
For anyone who ends up here who is using `jib` with gradle the following approach worked for me:
```
      "regexManagers": [
        {
          "description": "jib-base-image",
          "fileMatch": [
            "(^|/)gradle\\.properties$"
          ],
          "matchStringsStrategy": "any",
          "matchStrings": [
            "jibBaseImage=(?<depName>.*?):(?<currentValue>.*?)\\n"
          ],
          "datasourceTemplate": "docker"
          }
      ]
 ```
      
Where our `gradle.properties` file has the following snippet:

```
jibBaseImage=internal.artifactory.io/arch/ubi8-micro-java:17.0.6-jre-8.7-8-multi
```
We referenced the `jibBaseImage` when configuring jib plugin in `build.gradle.kts`:
```
val jibBaseImage: String by project

jib {
    from {
        image = jibBaseImage
        platforms {
            platform {
                architecture = "arm64"
                os = "linux"
            }
            platform {
                architecture = "amd64"
                os = "linux"
            }
        }
    }
}
```

Something to note:

I tried using the following match string:

```
jibBaseImage=(?<registryUrl>.*\\/)(?<depName>.*?):(?<currentValue>.*?)\\n
```
but since we're using a private artifactory registry for our docker images  it didn't seem work as the registryUrl bit wasn't producing the correct URLs to query artifactory for the image. I looked at how the regex was working when docker images were picked up from helm charts and followed that same pattern and lumping the registryUrl in with the depName seemed to resolve the problem.

**Reply by viceice** [reply]
this is a different issue, so please open a new discussion
